src/components/safe-apps/ApprovalEditor/ApprovalEditorForm.tsx

```diff
@@ -0,0 +1,107 @@
+import PrefixedEthHashInfo from '@/components/common/EthHashInfo'
+import NumberField from '@/components/common/NumberField'
+import TokenIcon from '@/components/common/TokenIcon'
+import { shortenAddress } from '@/utils/formatters'
+import { validateAmount, validateDecimalLength } from '@/utils/validation'
+import { Grid, Typography, Box, IconButton, SvgIcon } from '@mui/material'
+import { useState } from 'react'
+import { useForm } from 'react-hook-form'
+import css from './styles.module.css'
+import EditIcon from '@/public/images/common/edit.svg'
+import CheckIcon from '@mui/icons-material/Check'
+import type { ApprovalInfo } from '.'
+
+type FormData = {
+  approvals: string[]
+}
+
+export const ApprovalEditorForm = ({
+  approvalInfos,
+  updateApprovals,
+}: {
+  approvalInfos: ApprovalInfo[]
+  updateApprovals: (newApprovals: string[]) => void
+}) => {
+  const [editIDx, setEditIdx] = useState(-1)
+
+  const formMethods = useForm<FormData>({
+    defaultValues: {
+      approvals: approvalInfos.map((info) => info.amountFormatted),
+    },
+    mode: 'onChange',
+  })
+
+  const {
+    register,
+    formState: { errors },
+    getValues,
+  } = formMethods
+
+  const onSetEditing = (idx: number) => {
+    setEditIdx(idx)
+  }
+
+  const onSave = () => {
+    const formData = getValues('approvals')
+    updateApprovals(formData)
+    setEditIdx(-1)
+  }
+
+  return (
+    <Grid container direction="column">
+      {approvalInfos.map((tx, idx) => (
```

Comment:
You can simplify a lot of this logic by creating a "field array". RHF provides [`useFieldArray`](https://react-hook-form.com/api/usefieldarray/) specifically for this.

Reply:
I looked into it after the redesign and I don't think `useFieldArray` would change much of the logic. So I left it as is for now.